File: leetcode/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/Solution.java
import java.util.*;

class Solution {
    public List<List<Integer>> getAncestors(int n, int[][] edges) {
        List<List<Integer>> children = new ArrayList<>();
        for (int i = 0; i < n; i++) children.add(new ArrayList<>());
        int[] inDegree = new int[n];

        for (int[] e : edges) {
            children.get(e[0]).add(e[1]);
            inDegree[e[1]]++;
        }

        List<TreeSet<Integer>> ancestors = new ArrayList<>();
        for (int i = 0; i < n; i++) ancestors.add(new TreeSet<>());

        Deque<Integer> queue = new ArrayDeque<>();
        for (int v = 0; v < n; v++) {
            if (inDegree[v] == 0) queue.add(v);
        }

        while (!queue.isEmpty()) {
            int u = queue.poll();
            for (int v : children.get(u)) {
                ancestors.get(v).add(u);
                ancestors.get(v).addAll(ancestors.get(u));
                if (--inDegree[v] == 0) queue.add(v);
            }
        }

        List<List<Integer>> result = new ArrayList<>();
        for (TreeSet<Integer> s : ancestors) {
            result.add(new ArrayList<>(s));
        }
        return result;
    }
}

File: leetcode/2192-all-ancestors-of-a-node-in-a-directed-acyclic-graph/solution.cpp
#include <vector>
#include <set>
#include <queue>
using namespace std;

class Solution {
public:
    vector<vector<int>> getAncestors(int n, vector<vector<int>>& edges) {
        vector<vector<int>> children(n);
        vector<int> inDegree(n, 0);

        for (auto& e : edges) {
            children[e[0]].push_back(e[1]);
            inDegree[e[1]]++;
        }

        vector<set<int>> ancestors(n);

        queue<int> q;
        for (int v = 0; v < n; v++) {
            if (inDegree[v] == 0) q.push(v);
        }

        while (!q.empty()) {
            int u = q.front(); q.pop();
            for (int v : children[u]) {
                ancestors[v].insert(u);
                ancestors[v].insert(ancestors[u].begin(), ancestors[u].end());
                if (--inDegree[v] == 0) q.push(v);
            }
        }

        vector<vector<int>> result(n);
        for (int i = 0; i < n; i++) {
            result[i] = vector<int>(ancestors[i].begin(), ancestors[i].end());
        }
        return result;
    }
};
